spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,138 @@
+package org.apache.hadoop.fs.versioned
+
+import java.net.URI
+import java.util.Objects
+
+import com.gu.tableversions.metastore.VersionPaths
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+class VersionedFileSystem extends FileSystem {
+
+  private var baseURI: URI = _
+  private var baseFS: FileSystem = _
+  private var mappings: Map[String, String] = _
+  private var version: String = _
+
+  override def initialize(name: URI, conf: Configuration): Unit = {
+    val cacheDisabled = conf.getBoolean("fs.versioned.impl.disable.cache", false)
+    val baseFSScheme = conf.get("fs.versioned.baseFS")
+    version = conf.get("fs.versioned.version")
+
+    require(Objects.nonNull(baseFSScheme), "fs.versioned.baseFS not set in configuration")
+    require(Objects.nonNull(version), "fs.versioned.version not set in configuration")
+    require(cacheDisabled, "fs.versioned.impl.disable.cache not set to true in configuration")
+
+    import scala.collection.JavaConverters._
+
+    mappings = conf.iterator.asScala
+      .filter(_.getKey.startsWith(VersionPaths.KEY_PREFIX))
+      .map(_.getValue)
+      .toList
+      .map { v =>
+        (v, v + "/" + version)
+      }
+      .toMap
```

Comment:
Would probably make sense to put this in `VersionPaths` alongside the function that produces these config elements? They seem to logically belong together, and you could add a round-trip test as well then.

Reply:
Moved the other methods out of `VersionPaths` (as per other comment below). Might obsolete this change. 
Also, this would need revisiting if/when the proxy fs is changed to accomodate a different version for each partition.